source/simulator/core.cc

```diff
@@ -2393,72 +2393,32 @@ namespace aspect
+            if (pre_refinement_step < parameters.initial_adaptive_refinement)
+              {
+                do_initial_refinement_cycle(max_refinement_level);
+                ++pre_refinement_step;
+                goto start_time_iteration;
+              }
+            else
+              // invalidate the value of pre_refinement_step since it will no longer be used from here on
+              pre_refinement_step = std::numeric_limits<unsigned int>::max();
           }
 
-        // invalidate the value of pre_refinement_step since it will no longer be used from here on
-        if ( timestep_number == 0 )
-          pre_refinement_step = std::numeric_limits<unsigned int>::max();
-
         postprocess ();
 
-        // see if this is a time step where additional refinement is requested
-        // if so, then loop over as many times as this is necessary
-        if ((parameters.additional_refinement_times.size() > 0)
-            &&
-            (parameters.additional_refinement_times.front () < time+time_step))
-          {
-            while ((parameters.additional_refinement_times.size() > 0)
-                   &&
-                   (parameters.additional_refinement_times.front () < time+time_step))
-              {
-                ++max_refinement_level;
-                refine_mesh (max_refinement_level);
+        // see if this is a time step where refinement is requested
+        if (time_for_additional_refinement_cycle())
+          do_additional_refinement_cycle(max_refinement_level);
+        else if (time_for_regular_refinement_cycle())
+          refine_mesh (max_refinement_level);
 
-                parameters.additional_refinement_times
-                .erase (parameters.additional_refinement_times.begin());
-              }
-          }
-        else
-          // see if this is a time step where regular refinement is necessary, but only
-          // if the previous rule wasn't triggered
-          if (
-            (timestep_number > 0
-             &&
-             (parameters.adaptive_refinement_interval > 0)
-             &&
-             (timestep_number % parameters.adaptive_refinement_interval == 0))
-            ||
-            (timestep_number==0 && parameters.adaptive_refinement_interval == 1)
-          )
-            refine_mesh (max_refinement_level);
-
-        // every n time steps output a summary of the current
-        // timing information
-        if (((timestep_number > 0) && (parameters.timing_output_frequency != 0) &&
-             (timestep_number % parameters.timing_output_frequency == 0))
-            ||
-            (parameters.timing_output_frequency == 1)||(parameters.timing_output_frequency == 0))
-          {
-            computing_timer.print_summary ();
-            output_statistics();
-          }
+        // if requested output a summary of the current timing information
+        if (time_for_timing_output())
+          computing_timer.print_summary ();
```

Comment:
in other places, we have combined these two functions into a function that would be called `maybe_print_timing_output()` and that internally does the test and if necessary calls `print_summary()`. This would get you down to just one line here.